### PR TITLE
Added listener for disabled tests #4179

### LINKED ISSUE #4179: Intercept disabled tests
I would like to add to the report a list of tests that were disabled.

Is it possible to write an extension that intercepts disabled tests?

**BeforeEachListener** or **BeforeTestListener** callback won't be invoked if test is disabled.

### PR DESCRIPTION
**Pull request by kozlov-aa**
Closes #4179

### CODE REVIEW

**Comment by AlexCue987** [comment]
Very interesting. Can you elaborate on the problem(s) you are trying to solve, such as in which context this feature is going to help?

**Comment by kozlov-aa** [comment]
@AlexCue987 I am working on some automating testing project. We are using [Allure](https://allurereport.org/) as reporting tool. For sure we are integrate with allure by [kotest-extensions-allure](https://github.com/kotest/kotest-extensions-allure), but the report that this extension generates does not include ignored tests. This happen because [AllureTestReporter.kt](https://github.com/kotest/kotest-extensions-allure/blob/master/src/main/kotlin/io/kotest/extensions/allure/AllureTestReporter.kt) using **BeforeTestListener** and **AfterTestListener** hooks which does not invoke for ignored tests. However, if I am generate report by Junit native integration with Allure, report contains ignored tests, because Kotest tells Junit that test was ignored. This enhancement can help me to solve this problem. I just using new hook to inrich report by disabled tests.

File: kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TestEnabledCheckInterceptor.kt [hunk 1]
```diff
@@ -32,6 +35,7 @@ internal class TestEnabledCheckInterceptor(private val configuration: ProjectCon
          }
          false -> {
             logger.log { Pair(testCase.name.testName, "Test is disabled: ${enabled.reason}") }
+            testExtensions.disabledTestListenersInvocation(testCase, enabled.reason)
```
Comment (sksamuel) [inline kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TestEnabledCheckInterceptor.kt]: We prob want to fire this in the same place we fire the test engine listener, so we capture all places ?

Comment (flanker-c) [reply kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TestEnabledCheckInterceptor.kt]: I think yes, we want this.

Comment (flanker-c) [reply kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TestEnabledCheckInterceptor.kt]: So I think I should rename DisabledTestListener to IgnoredTestListener besause disable flags are not the only reason why a test might be ignored.

**Review by sksamuel — APPROVED** [review]

**Comment by kozlov-aa** [comment]
@sksamuel Forgot about :apiDump task.

**Review by sksamuel — APPROVED** [review]

File: kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/interceptors/EnabledCheckTestExecutionInterceptorTest.kt [hunk 2]
```diff
@@ -3,15 +3,19 @@ package com.sksamuel.kotest.engine.test.interceptors
...
 import io.kotest.core.test.TestCase
 import io.kotest.core.test.TestResult
 import io.kotest.core.test.TestType
-import io.kotest.engine.test.scopes.TerminalTestScope
 import io.kotest.engine.test.interceptors.TestEnabledCheckInterceptor
+import io.kotest.engine.test.scopes.TerminalTestScope
 import io.kotest.matchers.shouldBe
+import io.mockk.coVerify
+import io.mockk.every
+import io.mockk.spyk
```
Comment (OliverO2) [inline kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/interceptors/EnabledCheckTestExecutionInterceptorTest.kt:18]: These are not used, right?

File: kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/TestExtensions.kt [hunk 3]
```diff
@@ -151,4 +152,12 @@ internal class TestExtensions(private val registry: ExtensionRegistry) {
    fun logExtensions(testCase: TestCase): List<LogExtension> {
       return extensions(testCase).filterIsInstance<LogExtension>()
    }
+
+   /**
+    * Invokes all [IgnoredTestListener] for this test.
```
Comment (OliverO2) [inline kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/TestExtensions.kt]:
```suggestion
    * Invokes all [IgnoredTestListener]s for this test.
```

File: kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/extensions/test/IgnoredTestListenerTest.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+   val ignoredTestListener = spyk<IgnoredTestListener>()
+
+   register(ignoredTestListener)
+
+   describe("should be disabled by enabled flag in config").config(enabled = false) {
+      error("still enabled")
+   }
+
+   describe("!should be disabled by band") {
```
Comment (OliverO2) [inline kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/extensions/test/IgnoredTestListenerTest.kt]:
```suggestion
   describe("!should be disabled by bang") {
```

File: kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/extensions/test/IgnoredTestListenerTest.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+   describe("ignoredTestListener should be invoked") {
+      coVerify(exactly = 1) {
+         ignoredTestListener.ignoredTest(
+            testCase = match { it.name.testName == "should be disabled by enabled flag in config" },
+            reason = eq("Disabled by enabled flag in config")
+         )
+      }
+      coVerify(exactly = 1) {
+         ignoredTestListener.ignoredTest(
+            testCase = match { it.name.testName == "should be disabled by band" },
```
Comment (OliverO2) [inline kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/extensions/test/IgnoredTestListenerTest.kt]:
```suggestion
            testCase = match { it.name.testName == "should be disabled by bang" },
```

**Review by OliverO2 — COMMENTED** [review]
Some nitpicks.

I wonder if we could use tests without mocking? (Makes reviews and maintenance easier for those not familiar with mockk.)

File: kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/interceptors/EnabledCheckTestExecutionInterceptorTest.kt (hunk 2 above)
Comment (sksamuel) [reply kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/interceptors/EnabledCheckTestExecutionInterceptorTest.kt:18]: Removed all mockk from tests

**Comment by sksamuel** [comment]
I think this is breaking some behavior.

**Pull request merged**